tests/pubsub/testgossipmembership.nim

```diff
@@ -0,0 +1,176 @@
+    let topic = "test-topic"
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    # Subscribe to the topic
+    subscribeToTopics(gossipSub, @[topic])
+
+    # Check if the topic is present in the list of subscribed topics
+    check gossipSub.topics.contains(topic)
+
+    # Check if the topic is added to gossipsub and the peers list is not empty
+    check gossipSub.gossipsub[topic].len() > 0
+
+    # Close all peer connections and verify that they are properly cleaned up
+    await allFuturesThrowing(conns.mapIt(it.close()))
+
+    # Stop the gossipSub switch and wait for it to stop completely
+    await gossipSub.switch.stop()
+
+  # Simulate an UNSUBSCRIBE to the topic and check if the topic is removed from the relevant data structures but remains in gossipsub
+  asyncTest "handle UNSUBSCRIBE to the topic":
+    let topic = "test-topic"
+    let (gossipSub, conns) = setupGossipSub(topic, 5)
+
+    # Subscribe to the topic first
+    subscribeToTopics(gossipSub, @[topic])
+
+    # Now unsubscribe from the topic
+    unsubscribeFromTopics(gossipSub, @[topic])
+
+    # Verify the topic is removed from relevant structures
+    check topic notin gossipSub.topics
+    check topic notin gossipSub.mesh
+    check topic in gossipSub.gossipsub
+
+    # The topic should remain in gossipsub (for fanout)
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test subscribing and unsubscribing multiple topics
+  asyncTest "handle SUBSCRIBE and UNSUBSCRIBE multiple topics":
+    let topics = ["topic1", "topic2", "topic3"].toSeq()
+    let (gossipSub, conns) = setupGossipSub(topics, 5)
+
+    # Subscribe to multiple topics
+    subscribeToTopics(gossipSub, topics)
+
+    # Verify that all topics are added to the topics and gossipsub
+    check gossipSub.topics.len == 3
+    for topic in topics:
+      check gossipSub.gossipsub[topic].len() >= 0
+
+    # Unsubscribe from all topics
+    unsubscribeFromTopics(gossipSub, topics)
+
+    # Ensure topics are removed from topics and mesh, but still present in gossipsub
+    for topic in topics:
+      check topic notin gossipSub.topics
+      check topic notin gossipSub.mesh
+      check topic in gossipSub.gossipsub
+
+    await allFuturesThrowing(conns.mapIt(it.close()))
+    await gossipSub.switch.stop()
+
+  # Test ensuring that the number of subscriptions does not exceed the limit set in the GossipSub parameters
+  asyncTest "subscription limit test":
+    let gossipSub = TestGossipSub.init(newStandardSwitch())
+    gossipSub.topicsHigh = 10
+
+    var conns = newSeq[Connection]()
+    for i in 0 .. gossipSub.topicsHigh + 5:
+      let topic = "topic" & $i
+      # Ensure all topics are properly initialized before subscribing
+      gossipSub.mesh[topic] = initHashSet[PubSubPeer]()
+      gossipSub.topicParams[topic] = TopicParams.init()
+      gossipSub.gossipsub[topic] = initHashSet[PubSubPeer]()
+
+      if gossipSub.topics.len < gossipSub.topicsHigh:
```

Comment:
since we want to test the reaction of the logic if we subscribed to more than the limit then we need to try to add more subscribtions than the limit and check the reaction